File: js/formatters/ytTimestamp.ts
// For the Python implementation, see kuropy/formatters/ytTimestamp.py
/**
 * Parses timestamp retrieved from the YouTube API
 * 
 * @param ts A compatible timestamp
 */
export const ytTimestamp = (ts: string) => {
  const match = ts.match(/PT(\d+H)?(\d+M)?(\d+S)?/)

  if (!match) return "0:00"

  /* eslint-disable-next-line no-unused-vars */
  const [_, h, m, s] = match

  const _int = (n: string) => n ? parseInt(n, 10) : 0
  const _pad = (num: number) => num.toString().padStart(2, "0")
  const _concatTimeStr = (...val: (string | number)[]) => val.join(":")

  const hours = _int(h)
  const minutes = _int(m)
  const seconds = _int(s)

  const padMinute = _pad(minutes)
  const padSecond = _pad(seconds)

  const hourString = _concatTimeStr(hours, padMinute, padSecond)
  const minSecString = _concatTimeStr(padMinute, padSecond)

  return hours ? hourString : minSecString
}

File: kuropy/formatters/yt-timestamp.py
# For the JS implementation, see js/formatters/ytTimestamp.py
import re

__all__ = [
    "yt_timestamp"
]


def yt_timestamp(ts: str) -> str:
    """Parses timestamp retrieved from the YouTube API

    Args:
        ts (str): A compatible timestamp

    Returns:
        str: A parsed timecode
    """
    match = re.match(r"/PT(\d+H)?(\d+M)?(\d+S)?/", ts)

    if not match:
        return "0:00"

    _, h, m, s = match.groups()

    radix_int = lambda n: int(n[:-1]) if n else 0
    pad_str = lambda n: f"{n:02}"
    _concat_time_str = lambda *val: ":".join(map(str, val))

    hours = radix_int(h)
    minutes = radix_int(m)
    seconds = radix_int(s)

    pad_minute = pad_str(minutes)
    pad_second = pad_str(seconds)

    hour_string = _concat_time_str(hours, pad_minute, pad_second)
    min_sec_string = _concat_time_str(pad_minute, pad_second)

    return hour_string if hours else min_sec_string
